module/applications/dice/roll-configuration-dialog.mjs

```diff
@@ -0,0 +1,295 @@
+   * @param {BasicRollProcessConfiguration} config  Roll configuration data.
+   * @param {FormDataExtended} [formData]           Any data entered into the rolling prompt.
+   * @returns {BasicRoll[]}
+   * @private
+   */
+  #buildRolls(config, formData) {
+    const RollType = this.rollType;
+    return config.rolls?.map((config, index) => RollType.create(this._buildConfig(config, formData, index))) ?? [];
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Prepare individual configuration object before building a roll.
+   * @param {BasicRollConfiguration} config  Roll configuration data.
+   * @param {FormDataExtended} [formData]    Any data entered into the rolling prompt.
+   * @param {number} index                   Index of the roll within all rolls being prepared.
+   * @returns {BasicRollConfiguration}
+   * @protected
+   */
+  _buildConfig(config, formData, index) {
+    config = foundry.utils.mergeObject({ parts: [], data: {}, options: {} }, config);
+
+    /**
+     * A hook event that fires when a roll config is built using the roll prompt.
+     * @function dnd5e.buildRollConfig
+     * @memberof hookEvents
+     * @param {RollConfigurationDialog} app    Roll configuration dialog.
+     * @param {BasicRollConfiguration} config  Roll configuration data.
+     * @param {FormDataExtended} [formData]    Any data entered into the rolling prompt.
+     * @param {number} index                   Index of the roll within all rolls being prepared.
+     */
+    Hooks.callAll("dnd5e.buildRollConfig", this, config, formData, index);
+
+    if ( formData?.get("situational") && (config.situational !== false) ) {
+      config.parts.push("@situational");
+      config.data.situational = formData.get("situational");
+    }
+
+    return config;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Make any final modifications to rolls based on the button clicked.
+   * @param {string} action  Action on the button clicked.
+   * @returns {BasicRoll[]}
+   * @protected
+   */
+  _finalizeRolls(action) {
+    return this.rolls;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Rebuild rolls based on an updated config and re-render the dialog.
+   */
+  rebuild() {
+    this._onChangeForm(this.options.form, new Event("change"));
+  }
+
+  /* -------------------------------------------- */
+  /*  Event Handling                              */
+  /* -------------------------------------------- */
+
+  /**
+   * Handle submission of the dialog using the form buttons.
+   * @param {Event|SubmitEvent} event    The form submission event.
+   * @param {HTMLFormElement} form       The submitted form.
+   * @param {FormDataExtended} formData  Data from the dialog.
+   * @private
+   */
+  static async #handleFormSubmission(event, form, formData) {
+    const rolls = this._finalizeRolls(event.submitter?.dataset?.action);
+    await this.close({ dnd5e: { rolls } });
```

Comment:
We could avoid having to propagate these rolls via `options` if we change the contract for `_finalizeRolls` to expect subclasses to modify `this.rolls` in-place, rather than returning new rolls. That would mean that after `_finalizeRolls` is called, the dialog's `this.rolls` is the final version of the rolls, which is perhaps useful.

```suggestion
    this._finalizeRolls(event.submitter?.dataset?.action);
    this.close();
```

Reply:
Yeah, I can update that. Those main reason I was using closing options here is to know that the closing occurred via the submission versus occurring through the close button so I can know what to send to the `resolve` method.

Since the `close` method already takes `ApplicationClosingOptions`, it would be fantastic if core provided some more details indicating what triggered the closing.

Reply:
That is a good idea. Maybe we can add some properties to the close event.